### PR TITLE
Add Compilation Control doc

### PR DESCRIPTION
**Pull request by dsouzai**
* Move Object Lockword doc to features dir
* Use relative links in high level compiler README
* Move JIT Write Barries doc to concepts dir
* Move Best Practices dir to misc dir
* Add Compilation Control doc

### CODE REVIEW

**Comment by dsouzai** [comment]
@mpirvu @a7ehuo could you please review?

File: doc/compiler/control/CompilationControl.md [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+
+### -Xtune:virtualized
+
+This mode emphasises fast Startup and Rampup, and low CPU utilization,
+at the cost of a small throughput degradation. It is designed for
+use in virtualized environments.
+
+### -Xtune:throughput
+
+This mode emphases maximizing throughput performance.
```
Comment (dsouzai) [inline doc/compiler/control/CompilationControl.md:299]: @mpirvu Anything else I should add, i.e. tradeoffs, who it's designed for, etc.?

File: doc/compiler/control/CompilationControl.md [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+thread also has a Compilation Thread Monitor. This monitor is
+used to synchronize per thread information, such as its State,
+as well as to suspend and resume as needed.
+
+
+## Compilation Monitor
+
+The Compilation Monitor is a global monitor that is used to
+synchronize many aspects of the compilation infrastructure.
+
```
Comment (dsouzai) [inline doc/compiler/control/CompilationControl.md:138]: Is there anything more to add here? If not I can just remove this section altogether.

File: doc/compiler/concepts/JitWriteBarriers.md [hunk 3]
```diff
@@ -1,5 +1,5 @@
 <!--
-Copyright (c) 2021, 2021 IBM Corp. and others
+Copyright (c) 2021, 2022 IBM Corp. and others
```
Comment (mpirvu) [inline doc/compiler/concepts/JitWriteBarriers.md:2]: f there are no other changes, we should revert the year change

File: doc/compiler/control/CompilationControl.md [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+There are two main types of compilation; First Time Compilation and
+Recompilation. As the name suggests, First Time Compilations refer
+to the compilation of a method that has only been executed in the
+interpreter. Recompilations refer to methods that are recompiled,
+generally at a higher optimization level, to further improve
+performance.
+
+### Invocation Counts
+
+The JIT sets and invocation count for every method in every class
```
Comment (mpirvu) [inline doc/compiler/control/CompilationControl.md:50]: Typo: and --> an

File: doc/compiler/control/CompilationControl.md [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+that is loaded by the VM. This occurs in the JIT hook
+`jitHookInitializeSendTarget`. The count is stored in the `extra`
```
Comment (mpirvu) [inline doc/compiler/control/CompilationControl.md:52]: Maybe we should say that JNI methods do not undergo this process.

File: doc/compiler/control/CompilationControl.md [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+### Dynamic Loop Transfer
+
+Dynamic Loop Transfer (DLT) is a feature in OpenJ9 that allows
+a thread executing in the middle of an interpreted method to
+switch to the middle of a compiled method body and continue
+execution. This feature is used for methods that spend a lot of
+time in the interpreter but are not invoked many times; this
+happens when executing a long running loop.
+
+The JVM maintains an DLT buffer per Java Thread. Using
```
Comment (mpirvu) [inline doc/compiler/control/CompilationControl.md:85]: an --> a

File: doc/compiler/control/CompilationControl.md [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+run. Entries from the LPQ are only picked up by Compilation
+Threads when the main Compilation Queue is empty.
+
+
+## Compilation Threads
+
+Compilations are performed on dedicated threads known as Compilation
+Threads. These threads are created at JVM startup, and are
+suspended until there is work to be done. The exact number of
+threads created depends on the resources available on the machine,
```
Comment (mpirvu) [inline doc/compiler/control/CompilationControl.md:124]: on the resources available on the machine   --> on the CPU resources available to the JVM

Comment (mpirvu) [reply doc/compiler/control/CompilationControl.md:124]: The number of compilation threads created does not depend on the size of the compilation queue. The number of active compilation threads depends on the size of the compilation queue.

File: doc/compiler/control/CompilationControl.md [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+as well as the size of the Compilation Queue.
+
+In general, the Compilation Threads are synchronized using
+the [Compilation Monitor](#compilation-monitor). However, each
+thread also has a Compilation Thread Monitor. This monitor is
+used to synchronize per thread information, such as its State,
+as well as to suspend and resume as needed.
```
Comment (mpirvu) [inline doc/compiler/control/CompilationControl.md:131]: The Compilation Thread Monitor is also used for synchronization between a compilation thread and an application thread waiting for a compilation results from the said compilation thread.

File: doc/compiler/control/CompilationControl.md [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+
+## Compilation Monitor
+
+The Compilation Monitor is a global monitor that is used to
+synchronize many aspects of the compilation infrastructure.
+
+
+## Sampler Thread
+
+The Sampler Thread can be though of as a daemon within the JIT. It
```
Comment (mpirvu) [inline doc/compiler/control/CompilationControl.md:142]: though --> thought

File: doc/compiler/control/CompilationControl.md [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+
+### Async Checkpoint Sampling
+
+Async Checkpoints, also known as Safe Points and Yield Points, are
+well defined points during a Java Thread's execution wherein the
+thread yields to the VM for the purpose of doing GC or any other
+task that requires the thread to be stopped. Async Checkpoint Sampling
+piggybacks off this infrastructure to obtain sampling information
+that is used to determine which methods should be compiled and
+what what optimization level.
```
Comment (mpirvu) [inline doc/compiler/control/CompilationControl.md:154]: what what --> at what

File: doc/compiler/control/CompilationControl.md [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+
+This is the general procedure:
+
+1. Every 10ms, the Sampler Thread sets a flag in the J9VMThead of all Java Threads.
```
Comment (mpirvu) [inline doc/compiler/control/CompilationControl.md:158]: "of all Java Threads" --> "of all active Java Threads" (they need to have VM access)

File: doc/compiler/control/CompilationControl.md [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+2. At an Async Checkpoint, the VM checks if the flag has been set, and if so, calls a JIT hook.
```
Comment (mpirvu) [inline doc/compiler/control/CompilationControl.md:159]: It's not the VM that checks. Could be the JIT if the method runs jitted. I would rephrase: "When an application thread reaches an Async Checkpoint, it checks to see whether the flag has been set...".

File: doc/compiler/control/CompilationControl.md [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+
+
+## Code And Data Cache Management
+
+The compiled bodies are stored into the Code Cache; data related to the
+compiled body is stored in the Data Cache. Space in the Code and Data
+Caches have to be allocated on an as-needed basis. Additionally,
+these allocations have to be synchronized since multiple Compilation
+Threads can be active simultaneously. As such, each Compilation Thread,
+as part of the compilation, will reserve a Code Cache to ensure that
```
Comment (mpirvu) [inline doc/compiler/control/CompilationControl.md:175]: at the beginning of the codegen phase

File: doc/compiler/control/CompilationControl.md [hunk 14]
```diff
@@ -0,0 +1,304 @@
...
+is used. However, the actual chunk of memory is known as the "Code
+Respository", and individual subsections of the Code Repository are
+called "Code Caches". By default, the JIT allocates a 256 MiB Code
+Repository. This is then carved into 2 MiB Code Caches. It is this
+2 MiB subsection of the Code Reponsitory that a Compilation Thread
+reserves during a compilation.
+
+### Trampolines
+
+A trampolines is essentially a snippet of code that is used branch to
```
Comment (mpirvu) [inline doc/compiler/control/CompilationControl.md:189]: is used branch  -->  is used to branch

File: doc/compiler/control/CompilationControl.md [hunk 15]
```diff
@@ -0,0 +1,304 @@
...
+a location far from the current Instruction Pointer. Each Code Cache
```
Comment (mpirvu) [inline doc/compiler/control/CompilationControl.md:190]: The Trampoline section is not reserved if the JVM is configured such that it is guaranteed that far jumps will never be needed. This is the purpose of the code cache repository: to ensure that code caches are close together and obviate the need for  far jumps.

File: doc/compiler/control/CompilationControl.md [hunk 16]
```diff
@@ -0,0 +1,304 @@
...
+  - Decisions focus on maintaining throughput while minimizing CPU
+    consumed by the JIT (including the Sampler Thread).
+
+## Available Resources
+
+The JIT should minimize its impact on application performance.
+Therefore, how aggressively it can compile methods depends on
+the resources available to the JVM. Multiple compilation threads may
+only be activated if the system is not resource constrained.
+To minimize memory usage, only one higher optimization level
```
Comment (mpirvu) [inline doc/compiler/control/CompilationControl.md:238]: I vaguely remember that I performed some changes in this area to help Spark rampup faster. I don't recall what I did though.

File: doc/compiler/control/CompilationControl.md [hunk 17]
```diff
@@ -0,0 +1,304 @@
...
+
+The existence of the Shared Classes Cache (SCC) has a sigificant
+influence on the heuristics; this is because of SCC enables
+Ahead of Time (AOT) Compilation. See the [AOT docs](../aot) for
+technical details on the AOT feature.
+
+Because AOT compilations enable very fast startup, the JIT tries
+to maximize the number of compiled bodies it can load from the SCC
+(referred to as AOT Load). As a consequence, the invocation counts
+for AOT Compilations are lower than JIT Compilations. Additionally,
```
Comment (mpirvu) [inline doc/compiler/control/CompilationControl.md:264]: > As a consequence, the invocation counts for AOT Compilations are lower than JIT Compilations

This is not clear. It's the method that receives an invocation count, not the compilation. Moreover, we cannot know precisely which method will be AOTed.

File: doc/compiler/control/CompilationControl.md [hunk 18]
```diff
@@ -0,0 +1,304 @@
...
+the invocation counts for an AOT Load is even lower - by default
+it is 20.
+
+By very their nature, AOT method bodies have to be
```
Comment (mpirvu) [inline doc/compiler/control/CompilationControl.md:268]: By their very nature

File: doc/compiler/control/CompilationControl.md [hunk 19]
```diff
@@ -0,0 +1,304 @@
...
+relocated before they can be executed. As a consequence, the
+quality of AOT compiled code is, in general, less than that of JIT
```
Comment (mpirvu) [inline doc/compiler/control/CompilationControl.md:270]: Is the relocation the reason AOT has lower quality? I thought it's because the code has to be more generic to fit more than one JVM, whereas a regular JITed method is tailor fit for one JVM.

File: doc/compiler/control/CompilationControl.md (hunk 1 above)
Comment (mpirvu) [reply doc/compiler/control/CompilationControl.md:299]: emphases --> emphasizes

File: doc/compiler/control/CompilationControl.md [hunk 20]
```diff
@@ -0,0 +1,304 @@
...
+
+## Special Modes
+
+OpenJ9 has a few special flags that the user can specify to indicate
+to the JIT the kind of behaviour desired. This section outlines
+these modes.
+
+### -Xtune:virtualized
+
+This mode emphasises fast Startup and Rampup, and low CPU utilization,
```
Comment (mpirvu) [inline doc/compiler/control/CompilationControl.md:293]: emphasizes

File: doc/compiler/control/CompilationControl.md (hunk 1 above)
Comment (mpirvu) [reply doc/compiler/control/CompilationControl.md:299]: at the expense of other performance metrics like start-up time, footprint and CPU consumption

File: doc/compiler/control/CompilationControl.md [hunk 21]
```diff
@@ -0,0 +1,304 @@
...
+at the cost of a small throughput degradation. It is designed for
+use in virtualized environments.
+
+### -Xtune:throughput
+
+This mode emphases maximizing throughput performance.
+
+### -Xquickstart
+
+This mode emphasises minimizing time spent in Startup. It is designed
```
Comment (mpirvu) [inline doc/compiler/control/CompilationControl.md:303]: emphasizes quick start-up of applications

**Review by mpirvu — COMMENTED** [review]
I have a bunch of small comments

**Review by a7ehuo — COMMENTED** [review]
Could you add an entry for `ObjectLockword.md` in `doc/compiler/README.md`?

File: doc/compiler/control/CompilationControl.md (hunk 14 above)
Comment (a7ehuo) [inline doc/compiler/control/CompilationControl.md:189]: `A trampolines`
->
`A trampoline`

File: doc/compiler/control/CompilationControl.md [hunk 22]
```diff
@@ -0,0 +1,304 @@
...
+called "Code Caches". By default, the JIT allocates a 256 MiB Code
+Repository. This is then carved into 2 MiB Code Caches. It is this
+2 MiB subsection of the Code Reponsitory that a Compilation Thread
+reserves during a compilation.
+
+### Trampolines
+
+A trampolines is essentially a snippet of code that is used branch to
+a location far from the current Instruction Pointer. Each Code Cache
+has a section for storing Trampolines. This is so that methods in a
```
Comment (a7ehuo) [inline doc/compiler/control/CompilationControl.md:191]:
```suggestion
has a section for storing Trampolines so that methods in a
```

File: doc/compiler/control/CompilationControl.md [hunk 23]
```diff
@@ -0,0 +1,304 @@
...
+Collection, etc. These events can cause invalidations both in the
+generated code and the compilation. The former is addressed using
+[Runtime Assumptions](../runtime/RuntimeAssumption.md); the latter
+is addressed by aborting the compilation and possibly reattempting
+it.
+
+## Shared Classes Cache
+
+The existence of the Shared Classes Cache (SCC) has a sigificant
+influence on the heuristics; this is because of SCC enables
```
Comment (a7ehuo) [inline doc/compiler/control/CompilationControl.md:257]: `sigificant`
->
`significant`

`because of`
->
`because`

**Review by a7ehuo — COMMENTED** [review]
Thank you very much for creating this document @dsouzai! A few minor comments added

File: doc/compiler/control/CompilationControl.md [hunk 24]
```diff
@@ -0,0 +1,304 @@
...
+to the JIT the kind of behaviour desired. This section outlines
+these modes.
+
+### -Xtune:virtualized
+
+This mode emphasises fast Startup and Rampup, and low CPU utilization,
+at the cost of a small throughput degradation. It is designed for
+use in virtualized environments.
+
+### -Xtune:throughput
```
[the remark this answers is not in the document]
Comment (mpirvu) [reply doc/compiler/control/CompilationControl.md:297]: We decided not to talk about -Xtune:throughput just yet publicly. There are some instances when throughput is actually degraded by this option.

File: doc/compiler/misc/BestPractices.md [hunk 25]
```diff
@@ -1,5 +1,5 @@
 <!--
-Copyright (c) 2019, 2020 IBM Corp. and others
+Copyright (c) 2019, 2022 IBM Corp. and others
```
[the remark this answers is not in the document]
Comment (dsouzai) [reply doc/compiler/misc/BestPractices.md:2]: I think the copyright checker will complain because it just checks for what files were affected in a PR. Also technically the copyrights of all files should be changed every year, but we change it on an as needed basis to prevent a yearly massive CR change.

File: doc/compiler/control/CompilationControl.md [hunk 26]
```diff
@@ -0,0 +1,304 @@
...
+
+### Invocation Counts
+
+The JIT sets and invocation count for every method in every class
+that is loaded by the VM. This occurs in the JIT hook
+`jitHookInitializeSendTarget`. The count is stored in the `extra`
+field of the `J9Method`. This field is also low-tagged if the
+method is interpreted. As such the count is encoded by left-shifting
+1 bit.
+
```
[the remark this answers is not in the document]
Comment (dsouzai) [reply doc/compiler/control/CompilationControl.md:57]: I added the bit about decoding. However, I'll skip on the example because I think it's too much detail for an overview doc.

**Comment by dsouzai** [comment]
@mpirvu @a7ehuo ready for another pass. see force pushes [one](https://github.com/eclipse-openj9/openj9/compare/05bd9672821a16adb4289e7359cca79db3ccb8a0..986ab385cf514581bddeaadf5f331a7b13ea2613) and [two](https://github.com/eclipse-openj9/openj9/compare/986ab385cf514581bddeaadf5f331a7b13ea2613..d02c910dcfb4b4cb946c73457e3f0cb10ad99e14)

**Review by a7ehuo — APPROVED** [review]
LGTM

File: doc/compiler/README.md [hunk 27]
```diff
@@ -164,7 +165,7 @@ In the end code generators perform binary encoding to write the appropriate bits
 * <details><summary><b>8. Memory</b></summary>
 
   * [OMR Compiler Memory Manager (OMR)](https://github.com/eclipse/omr/blob/master/doc/compiler/memory/MemoryManager.md)
-  * [OpenJ9 Compiler Memory Manager (OpenJ9)](https://github.com/eclipse-openj9/openj9/blob/master/doc/compiler/memory/MemoryManager.md)
+  * [OpenJ9 Compiler Memory Manager (OpenJ9)](memory/MemoryManager.md)
```
Comment (hzongaro) [inline doc/compiler/README.md:168]: There are two more lingering absolute URIs below:

```
* <b>9. [AOT](aot/README.md)</b>
* <b>10. [JITServer](jitserver/README.md)</b>
```

**Comment by dsouzai** [comment]
@hzongaro made requested edits, see [force push](https://github.com/eclipse-openj9/openj9/compare/d02c910dcfb4b4cb946c73457e3f0cb10ad99e14..c2ec6150dac711890982a7236ab994458e1bbab9).

File: doc/compiler/control/CompilationControl.md [hunk 28]
```diff
@@ -0,0 +1,325 @@
...
+[Runtime Assumptions](../runtime/RuntimeAssumption.md); the latter
+is addressed by aborting the compilation and possibly reattempting
+it.
+
+## Shared Classes Cache
+
+The existence of the Shared Classes Cache (SCC) has a significant
+influence on the heuristics; this is because the SCC allows for
+Ahead of Time (AOT) Compilation. See the [AOT docs](../aot) for
+technical details on the AOT feature.
```
[the remark this answers is not in the document]
Comment (dsouzai) [reply doc/compiler/control/CompilationControl.md:302]: Yeah I think that's something that influences the JIT's heuristics; the profiling information in the SCC doesn't affect the Control infra's heuristics; at least I don't think it does..

File: doc/compiler/control/CompilationControl.md [hunk 29]
```diff
@@ -0,0 +1,325 @@
...
+of the codegen phase to ensure that no other thread can write to it.
+
+There unfortunately exists a bit of name overloading. When speaking
+generally about a place to store compiled bodies, the term "Code Cache"
+is used. However, the actual chunk of memory is known as the "Code
+Repository", and individual subsections of the Code Repository are
+called "Code Caches". By default, the JIT allocates a 256 MiB Code
+Repository. This is then carved into 2 MiB Code Caches. It is this
+2 MiB subsection of the Code Repository that a Compilation Thread
+reserves during a compilation.
```
[the remark this answers is not in the document]
Comment (dsouzai) [reply doc/compiler/control/CompilationControl.md]: Heh there seems to be a complicated set of conditions for the sizes set:
https://github.com/eclipse-openj9/openj9/blob/78a2268e24f5562bb30daea946477669f300b323/runtime/compiler/control/rossa.cpp#L1181-L1209
So I think for the sake of simplicity I'm just gonna info about the size of the data cache.

File: doc/compiler/control/CompilationControl.md [hunk 30]
```diff
@@ -0,0 +1,354 @@
...
+  - Decisions focus on maintaining throughput.
+- DEEPSTEADY_STATE
+  - Decisions focus on maintaining throughput while minimizing CPU
+    consumed by the JIT (including the Sampler Thread).
+
+Startup refers to the period of time between the start of the JVM and
+the point at which the Java Application is ready for load. Rampup
+refers to the period of time between the end of Startup and the point
+at which the Java Application is running at peak throughput. When the
+application runs at peak throughput, it is said to be in Steady State.
```
Comment (0xdaryl) [inline doc/compiler/control/CompilationControl.md:273]: I feel each of these sentences should move up to their corresponding state description.

**Review by 0xdaryl — COMMENTED** [review]
Great overall introduction to compilation control!  Thanks for writing this up.

File: doc/compiler/control/CompilationControl.md [hunk 31]
```diff
@@ -0,0 +1,354 @@
...
+
+The compiled bodies are stored into the Code Repository; data related
+to the compiled body is stored in Data Caches. An example of such data
+is the `J9JITExceptionTable` (also `typedef`'d as `TR_MetaData`) that
+is used to describe the body (e.g. start and end address, optimization
+level, etc.). Space in the Code Repository and Data Caches have to be
+allocated on an as-needed basis. Additionally, these allocations have
+to be synchronized since multiple Compilation Threads can be active
+simultaneously. As such, each Compilation Thread, as part of the
+compilation, will reserve a Code and Data Cache at the beginning of the
```
[the remark this answers is not in the document]
Comment (dsouzai) [reply doc/compiler/control/CompilationControl.md:213]: Yes you're right, we only reserve the data cache for tosscode, AOT, and jitserver.

**Review by mpirvu — APPROVED** [review]

**Comment by mpirvu** [comment]
All comments from reviewers are addressed. This PR is ready to be merged.
